conans/client/proxy.py

```diff
@@ -143,6 +143,14 @@ def _refresh():
 
         if os.path.exists(conanfile_path):
             log_recipe_got_from_local_cache(conan_reference)
+
+            # If we have more than one defined remote and for some reason we do not know where
+            #  this ref came from, we should search in the remotes again for this recipe so we
+            #  can properly find it later.  Otherwise we assume the first remote in the list.
+            #  (Which may be wrong.)
+            if len(self._registry.remotes) > 1 and not self._registry.get_ref(conan_reference):
```

Comment:
I don't think this is doing something desired. If you are creating packages, locally in your computer, the recipe you are developing with your latest changes will be overwritten by the upstream, older one.

The local cache has precedence over the remote, this is part of the decentralized paradigm.

Reply:
Is there an alternate means for identifying the upstream without actually downloading the recipe.  It looked like the logic always used the successful download.

Reply:
It is not that the upstream can be automatically identified. If you start with a clean, fresh installation of conan, and you do some ``conan create Pkg/0.1@user/channel``, you don't want to automatically assign a remote to it, just because a package with the same reference exists for example in the conan-center repository. 

Conan tries to follows the model of decentralization of ``git``. So if you are operating locally, you have absolutely no idea about the possible remotes, and they wouldn't be even check at all. The same as you can do git init, add, commit, even clone, totally locally, you can ``conan create`` packages.

At the moment you explicitly want, you can assign a remote. In ``conan upload`` such remote is assigned to the package, the same as with ``git push`` in which the remote branch is followed. The same as in git, in which you can set a remote branch without needing to push it, in conan you can ``conan remote add_ref`` to explicitly set a remote for a package.

Reply:
Probably the issue is:

> conan data cache is shared into a docker container

Are you sharing just the ``.conan/data`` folder? The map between packages and remotes is stored in ``.conan/registry.txt``, so probably you would like to share the whole ``.conan`` folder to docker?

Reply:
I see your point about local recipes.

I am sharing just the `.conan/data` directory because sometimes the docker containers are different Linux distributions with different gcc versions and such.  I can toy with just sharing the `.conan/registry.txt` file, but the remote definitions may also be different between containers.

Reply:
Oh, that is an atypical setup, as conan servers can host artifacts for all OS and platforms, typically the remotes are always the same. If you have different configurations, you might be interested in the feature we are releasing in the next version: ``conan config install`` that allows to sync configuration (settings.yml, remotes, conan.conf, profiles) with an external origin. This is the PR if you want to follow or try it: https://github.com/conan-io/conan/pull/1705

Reply:
We have it set up like this so only the secure builders can push to one repo, and all devs can push to the other.

I will take a look at the config install and see if it will work in our environment.  Thanks!

Reply:
If you really want to make sure that only some "secure builders" (typically CI) can push to one repo, the typical thing is to use a different username for the package, and then control permissions in the servers. The approach most users follow is that CI will be creating packages like ``Pkg/0.1@teamname/stable``, while developers create packages like ``Pkg/0.1@memsharded/stable``, and permissions are controlled to allow ``*teamname*`` packages. 

> I will take a look at the config install and see if it will work in our environment. Thanks!

Good. Now you need to run it from sources from my branch, but we might try to release it in 2-3 weeks, so keep tuned :)

Reply:
We are using Artifactory for the server backend, which does not seem to enforce the user credentials the same way.  It lets me push as any user name.  (Maybe that is an artifactory config, but I am not in control of that to know.)

Reply:
I am not an expert in Artifactory, but I think it is possible to configure permissions for this kind of things. Maybe you or people in control might want to check it with Artifactory support. Thanks for all the feedback!

Reply:
Will do, thanks for the help.